### PR TITLE
Update REST Assessment Cheat Sheet: OpenAPI, JWT/OAuth2, BOLA, Mass Assignment, Rate Limiting

### LINKED ISSUE #2023: Update: REST Assessment Cheat Sheet – Modern OpenAPI & JWT Coverage
## What is missing or needs to be updated?

The current REST Assessment Cheat Sheet provides strong foundational guidance for assessing RESTful services. However, modern REST APIs have evolved significantly and are now commonly:

- OpenAPI / Swagger-driven  
- JWT/OAuth2 secured  
- Cloud-native and async  
- Built using frameworks such as FastAPI, Express, Django REST Framework, etc.

The cheat sheet references WSDL/WADL for service description but does not address modern OpenAPI-based attack surface discovery or schema-driven testing approaches.

In addition, several API-specific vulnerabilities that are highly relevant today are not explicitly covered, including:

- Broken Object Level Authorization (BOLA / IDOR in APIs)
- Mass Assignment vulnerabilities in JSON APIs
- JWT signature validation issues
- OAuth2 scope and role misconfigurations
- Rate limiting and throttling assessment
- Schema-based fuzzing using OpenAPI definitions

Since the OWASP API Security Top 10 is now widely adopted, aligning this cheat sheet more closely with modern API assessment practices would make it more relevant for current testing scenarios.

---

## How should this be resolved?

The cheat sheet could be enhanced by adding dedicated sections covering the following areas:

### 1. Assessing OpenAPI/Swagger-based REST APIs
- Enumerating endpoints via `/openapi.json`, `/swagger.json`, or `/docs`
- Leveraging schema definitions for structured fuzzing
- Testing undocumented or hidden fields

### 2. JWT and OAuth2 Assessment
- Verifying proper signature validation
- Testing expired, malformed, or tampered tokens
- Validating scope and role enforcement

### 3. Broken Object Level Authorization (BOLA)
- IDOR testing patterns in REST endpoints
- Horizontal and vertical privilege escalation scenarios

### 4. Mass Assignment in JSON APIs
- Detecting unintended field binding
- Testing for privilege escalation via request body manipulation

### 5. Rate Limiting & Throttling Assessment
- Identifying missing or weak rate limiting
- Evaluating burst and abuse handling mechanisms

I would be happy to draft and contribute these sections if this direction aligns with the maintainers’ vision for the cheat sheet.

### PR DESCRIPTION
**Pull request by franrob-projects**
## What this PR addresses

Closes #2023

The REST Assessment Cheat Sheet is a solid foundation but references WSDL/WADL as service description formats without addressing OpenAPI — the standard that now describes virtually every modern REST API. It also has no coverage of the authentication mechanisms, authorization flaws, or abuse vectors most commonly found in APIs today.

## Changes

Five new sections added between the existing fuzzing guidance and the Related Resources footer:

### 1. Assessing OpenAPI/Swagger-based REST APIs
- Common schema discovery paths (`/openapi.json`, `/swagger.json`, `/docs`, etc.)
- Using the schema for complete endpoint enumeration, including paths not exposed in the UI
- Undocumented field discovery
- Schema-driven fuzzing with ZAP and Schemathesis

### 2. JWT and OAuth2 Assessment
- `alg: none` and algorithm confusion attacks
- Claim tampering and expiry bypass
- `kid` header injection (SQL injection, path traversal, SSRF)
- OAuth2: scope enforcement, cross-tenant token reuse, PKCE requirement, refresh token rotation

### 3. Broken Object Level Authorization (BOLA)
- Horizontal and vertical privilege escalation patterns
- Testing all HTTP methods independently
- Indirect/encoded reference enumeration

### 4. Mass Assignment in JSON APIs
- Identifying bindable fields from schema/response comparison
- Injecting extra fields in POST, PUT, and PATCH requests
- Numeric privilege escalation targets (balance, credits, role)

### 5. Rate Limiting and Throttling Assessment
- Baseline measurement and standard rate-limit headers
- Confirming `429` enforcement
- Bypass techniques (IP header rotation, method switching)
- Sensitive operation coverage (OTP, password reset, enumeration)

Updated Related Resources section to include the OWASP API Security Top 10, JWT Cheat Sheet, OAuth2 Cheat Sheet, and Schemathesis.

## Checklist

- [x] All markdown files do not raise any validation policy violation
- [x] All markdown files follow the format rules
- [x] Any references to websites have been formatted as `[TEXT](URL)`
- [x] The contribution has been verified for accuracy against current attack/defence literature

## AI Tool Usage Disclosure

- [ ] I have not used AI tools to generate the contents of this PR

### CODE REVIEW

**Review by andrzejsydor — APPROVED** [review]

File: cheatsheets/REST_Assessment_Cheat_Sheet.md [hunk 1]
```diff
@@ -62,7 +62,129 @@ Analyzing collected requests to optimize [fuzzing](https://owasp.org/www-communi
...
+
+**Schema-driven fuzzing.** Tools such as [OWASP ZAP](https://www.zaproxy.org/) and [Schemathesis](https://schemathesis.readthedocs.io/) can import an OpenAPI definition and automatically generate test cases for each endpoint, including boundary values, type mismatches, and required-field omissions.
+
+## JWT and OAuth2 Assessment
+
+Many modern REST APIs use [JSON Web Tokens (JWTs)](https://jwt.io/) for stateless authentication and [OAuth 2.0](https://oauth.net/2/) for delegated authorization. Misconfigurations in either are a frequent source of critical vulnerabilities.
+
+### JWT Assessment
+
+**Verify signature validation.** Swap the algorithm header to `alg: none` and remove the signature. A vulnerable server will accept the token without validating it.
```
Comment (mackowski) [inline cheatsheets/REST_Assessment_Cheat_Sheet.md:94]: The actual attack retains the trailing . — the token structure must remain header.payload. (with an empty signature segment), not     
  header.payload. Some JWT libraries reject a token that is missing the final period entirely. The description should read:
                                                                                                                                        
  ▎ Strip the signature body but keep the trailing period: header.payload.

File: cheatsheets/REST_Assessment_Cheat_Sheet.md
Comment (mackowski) [inline cheatsheets/REST_Assessment_Cheat_Sheet.md:17]: The new OpenAPI section says OpenAPI "replaces older formats such as WSDL and WADL," but the original text is left untouched. Readers 
  will encounter the old claim and the new framing in the same document without reconciliation. The existing paragraph should be updated
   to acknowledge OpenAPI as the current standard.

**Review by mackowski — CHANGES_REQUESTED** [review]
Please fix two comment and linter errors

**Comment by mackowski** [comment]
@franrob-projects any updates on this PR?

File: cheatsheets/REST_Assessment_Cheat_Sheet.md [hunk 2]
```diff
@@ -62,7 +62,129 @@ Analyzing collected requests to optimize [fuzzing](https://owasp.org/www-communi
...
+
+**Test algorithm confusion.** If the server uses RS256 (asymmetric), try signing the token with HS256 using the server's *public key* as the HMAC secret. Servers that do not pin the expected algorithm may accept the forged token.
+
+**Tamper with claims.** Decode the payload (it is Base64Url-encoded, not encrypted), modify claims such as `sub`, `role`, or `email`, re-sign or strip the signature, and replay the token.
+
+**Check expiry enforcement.** Send a token whose `exp` claim is in the past. The server should return `401`; if it returns `200`, expiry is not being validated.
+
+**Inspect the payload for sensitive data.** JWT payloads are readable by anyone who holds the token. Look for PII, internal identifiers, or role information that should not be client-visible.
```
Comment (randomstuff) [inline cheatsheets/REST_Assessment_Cheat_Sheet.md:102]: > JWT payloads are readable by anyone who holds the token

Maybe add something like "by default" (because encrypted JWT is a thing)

File: cheatsheets/REST_Assessment_Cheat_Sheet.md [hunk 3]
```diff
@@ -62,7 +62,129 @@ Analyzing collected requests to optimize [fuzzing](https://owasp.org/www-communi
...
+
+For comprehensive JWT attack guidance see the [OWASP JSON Web Token Cheat Sheet](JSON_Web_Token_for_Java_Cheat_Sheet.md).
+
+### OAuth2 Assessment
+
+**Validate scope enforcement.** Obtain a token with a limited scope (e.g. `read:profile`) and attempt operations that require broader scope (e.g. `write:admin`). The server should reject out-of-scope requests with `403`.
+
+**Test token reuse across tenants.** In multi-tenant APIs, use a valid token issued to Tenant A and attempt to access resources belonging to Tenant B.
+
+**Check PKCE enforcement.** For authorization code flows, confirm that [PKCE](https://oauth.net/2/pkce/) (`code_challenge` / `code_verifier`) is required. Without it, a stolen authorization code can be exchanged for a token.
```
Comment (randomstuff) [inline cheatsheets/REST_Assessment_Cheat_Sheet.md:114]: > Without it, a stolen authorization code can be exchanged for a token.

This assume a public client or stolen client credentials.

File: cheatsheets/REST_Assessment_Cheat_Sheet.md [hunk 4]
```diff
@@ -62,7 +62,129 @@ Analyzing collected requests to optimize [fuzzing](https://owasp.org/www-communi
...
+- `X-RateLimit-Reset` — Unix timestamp when the window resets
+- `Retry-After` — seconds to wait after a `429` response
+
+**Exceed the limit.** Send enough requests to exhaust the documented limit and confirm the API returns `429 Too Many Requests`. A `200` response after the limit is exceeded indicates rate limiting is not enforced.
+
+**Test per-resource limits separately.** Authentication endpoints (login, password reset, OTP verification) should have stricter limits than general data endpoints. Test each independently.
+
+**Attempt limit bypass.** Try the following bypass techniques and observe whether the limit resets or is circumvented:
+
+- Rotating `X-Forwarded-For` or `X-Real-IP` header values.
```
Comment (randomstuff) [inline cheatsheets/REST_Assessment_Cheat_Sheet.md:177]: Maybe the [`Forwarded` header](https://datatracker.ietf.org/doc/html/rfc7239) as well. Anyone actually using this header?

**Comment by mackowski** [comment]
I am closing this PR because of inactivity. @franrob-projects feel free to fix and re-open it
